### PR TITLE
[SQL MI]Add support for Service principal and guest users in Set-AzSqlInstanceActiveDirectoryAdministrator cmdlet

### PR DESCRIPTION
**Pull request by amolagar5**
Added support for Service principal and guest users in Set-AzSqlInstanceActiveDirectoryAdministrator cmdlet

## Checklist

- [x] I have read the [_Submitting Changes_](../blob/master/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/master/CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [x] The appropriate `ChangeLog.md` file(s) has been updated:
    - For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    - A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header -- no new version header should be added
- [x] The PR does not introduce [breaking changes](../blob/master/documentation/breaking-changes/breaking-changes-definition.md)
- [ ] If applicable, the changes made in the PR have proper test coverage
- [ ] For public API changes to cmdlets:
    - [ ] a cmdlet design review was approved for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (_Microsoft internal only_)
    - [ ] the markdown help files have been regenerated using the commands listed [here](../blob/master/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)

### CODE REVIEW

**Comment by adxsdkps** [comment]
Can one of the admins verify this patch?

**Comment by VeryEarly** [comment]
Updated changelog.md

please update ## Upcoming release session only

File: src/Sql/Sql/InstanceActiveDirectoryAdministrator/Services/AzureSqlInstanceActiveDirectoryAdministratorAdapter.cs [hunk 1]
```diff
@@ -173,31 +176,77 @@ protected ManagedInstanceAdministrator GetActiveDirectoryInformation(string disp
...
+            {
+                var applicationIdString = objectId.ToString();
+                odataQueryFilter = new Rest.Azure.OData.ODataQuery<ServicePrincipal>(a => a.AppId == applicationIdString);
+            }
+            else
+            {
+                odataQueryFilter = new Rest.Azure.OData.ODataQuery<ServicePrincipal>(a => a.DisplayName == displayName);
+            }
+
+            var srevicePrincipalList = ActiveDirectoryClient.FilterServicePrincipals(odataQueryFilter);
```
Comment (dmitarredz) [inline src/Sql/Sql/InstanceActiveDirectoryAdministrator/Services/AzureSqlInstanceActiveDirectoryAdministratorAdapter.cs:209]: nit: servicePrincipalList

**Comment by amolagar5** [comment]
> Hey @VeryEarly ,
> will fix changeLog
> can you please help me with tests. Exiting tests fail with this error. Manually scenario is working. Do we need to set some environment variable for this.
> System.Management.Automation.ActionPreferenceStopException : The running command stopped because the preference variable "ErrorActionPreference" or common parameter is set to Stop: Unable to find a matching HTTP request for URL 'GET /0c1edf5d-e5c5-4aca-ab69-ef194134f44b/servicePrincipals?$filter=appId%20eq%20'52b6d571-5ff9-4b8f-92de-4a5b1bcdbbef'&api-version=1.6'. Calling method Item().
> Stack Trace:
> at System.Management.Automation.Runspaces.PipelineBase.Invoke(IEnumerable input)
> at System.Management.Automation.PowerShell.Worker.ConstructPipelineAndDoWork(Runspace rs, Boolean performSyncInvoke)
> at System.Management.Automation.PowerShell.Worker.CreateRunspaceIfNeededAndDoWork(Runspace rsToUse, Boolean isSync)
> at System.Management.Automation.PowerShell.CoreInvokeHelper[TInput,TOutput](PSDataCollection`1 input, PSDataCollection`1 output, PSInvocationSettings settings)
> at System.Management.Automation.PowerShell.CoreInvoke[TInput,TOutput](PSDataCollection`1 input, PSDataCollection`1 output, PSInvocationSettings settings)
> at System.Management.Automation.PowerShell.CoreInvoke[TOutput](IEnumerable input, PSDataCollection`1 output, PSInvocationSettings settings)
> at System.Management.Automation.PowerShell.Invoke(IEnumerable input, PSInvocationSettings settings)
> at System.Management.Automation.PowerShell.Invoke()
> at Microsoft.WindowsAzure.Commands.ScenarioTest.EnvironmentSetupHelper.RunPowerShellTest(String[] scripts) in D:\a\1\s\tools\ScenarioTest.ResourceManager\EnvironmentSetupHelper.cs:line 571
> at Microsoft.Azure.Commands.ScenarioTest.SqlTests.SqlTestsBase.RunPowerShellTest(String[] scripts) in D:\a\1\s\src\Sql\Sql.Test\ScenarioTests\SqlTestsBase.cs:line 103
> at Microsoft.Azure.Commands.Sql.Test.ScenarioTests.ManagedInstanceActiveDirectoryAdministratorTests.TestManagedInstanceActiveDirectoryAdministrator() in D:\a\1\s\src\Sql\Sql.Test\ScenarioTests\ManagedInstance

@msJinLei @dingmeng-xue any idea about this. looks like tests scenarios not working with service principals.

**Comment by VeryEarly** [comment]
> > Hey @VeryEarly ,
> > will fix changeLog
> > can you please help me with tests. Exiting tests fail with this error. Manually scenario is working. Do we need to set some environment variable for this.
> > System.Management.Automation.ActionPreferenceStopException : The running command stopped because the preference variable "ErrorActionPreference" or common parameter is set to Stop: Unable to find a matching HTTP request for URL 'GET /0c1edf5d-e5c5-4aca-ab69-ef194134f44b/servicePrincipals?$filter=appId%20eq%20'52b6d571-5ff9-4b8f-92de-4a5b1bcdbbef'&api-version=1.6'. Calling method Item().
> > Stack Trace:
> > at System.Management.Automation.Runspaces.PipelineBase.Invoke(IEnumerable input)
> > at System.Management.Automation.PowerShell.Worker.ConstructPipelineAndDoWork(Runspace rs, Boolean performSyncInvoke)
> > at System.Management.Automation.PowerShell.Worker.CreateRunspaceIfNeededAndDoWork(Runspace rsToUse, Boolean isSync)
> > at System.Management.Automation.PowerShell.CoreInvokeHelper[TInput,TOutput](PSDataCollection`1 input, PSDataCollection`1 output, PSInvocationSettings settings)
> > at System.Management.Automation.PowerShell.CoreInvoke[TInput,TOutput](PSDataCollection`1 input, PSDataCollection`1 output, PSInvocationSettings settings)
> > at System.Management.Automation.PowerShell.CoreInvoke[TOutput](IEnumerable input, PSDataCollection`1 output, PSInvocationSettings settings)
> > at System.Management.Automation.PowerShell.Invoke(IEnumerable input, PSInvocationSettings settings)
> > at System.Management.Automation.PowerShell.Invoke()
> > at Microsoft.WindowsAzure.Commands.ScenarioTest.EnvironmentSetupHelper.RunPowerShellTest(String[] scripts) in D:\a\1\s\tools\ScenarioTest.ResourceManager\EnvironmentSetupHelper.cs:line 571
> > at Microsoft.Azure.Commands.ScenarioTest.SqlTests.SqlTestsBase.RunPowerShellTest(String[] scripts) in D:\a\1\s\src\Sql\Sql.Test\ScenarioTests\SqlTestsBase.cs:line 103
> > at Microsoft.Azure.Commands.Sql.Test.ScenarioTests.ManagedInstanceActiveDirectoryAdministratorTests.TestManagedInstanceActiveDirectoryAdministrator() in D:\a\1\s\src\Sql\Sql.Test\ScenarioTests\ManagedInstance
> 
> @msJinLei @dingmeng-xue any idea about this. looks like tests scenarios not working with service principals.

In your updated code, there are API calls to get serviceprincipal, so please re-record testcase: https://github.com/Azure/azure-powershell/blob/amagarwa/dev/sqlmiaddSPGuestuserSupport/src/Sql/Sql.Test/ScenarioTests/ManagedInstanceActiveDirectoryAdministratorTests.ps1

**Comment by amolagar5** [comment]
> > > Hey @VeryEarly ,
> > > will fix changeLog
> > > can you please help me with tests. Exiting tests fail with this error. Manually scenario is working. Do we need to set some environment variable for this.
> > > System.Management.Automation.ActionPreferenceStopException : The running command stopped because the preference variable "ErrorActionPreference" or common parameter is set to Stop: Unable to find a matching HTTP request for URL 'GET /0c1edf5d-e5c5-4aca-ab69-ef194134f44b/servicePrincipals?$filter=appId%20eq%20'52b6d571-5ff9-4b8f-92de-4a5b1bcdbbef'&api-version=1.6'. Calling method Item().
> > > Stack Trace:
> > > at System.Management.Automation.Runspaces.PipelineBase.Invoke(IEnumerable input)
> > > at System.Management.Automation.PowerShell.Worker.ConstructPipelineAndDoWork(Runspace rs, Boolean performSyncInvoke)
> > > at System.Management.Automation.PowerShell.Worker.CreateRunspaceIfNeededAndDoWork(Runspace rsToUse, Boolean isSync)
> > > at System.Management.Automation.PowerShell.CoreInvokeHelper[TInput,TOutput](PSDataCollection`1 input, PSDataCollection`1 output, PSInvocationSettings settings)
> > > at System.Management.Automation.PowerShell.CoreInvoke[TInput,TOutput](PSDataCollection`1 input, PSDataCollection`1 output, PSInvocationSettings settings)
> > > at System.Management.Automation.PowerShell.CoreInvoke[TOutput](IEnumerable input, PSDataCollection`1 output, PSInvocationSettings settings)
> > > at System.Management.Automation.PowerShell.Invoke(IEnumerable input, PSInvocationSettings settings)
> > > at System.Management.Automation.PowerShell.Invoke()
> > > at Microsoft.WindowsAzure.Commands.ScenarioTest.EnvironmentSetupHelper.RunPowerShellTest(String[] scripts) in D:\a\1\s\tools\ScenarioTest.ResourceManager\EnvironmentSetupHelper.cs:line 571
> > > at Microsoft.Azure.Commands.ScenarioTest.SqlTests.SqlTestsBase.RunPowerShellTest(String[] scripts) in D:\a\1\s\src\Sql\Sql.Test\ScenarioTests\SqlTestsBase.cs:line 103
> > > at Microsoft.Azure.Commands.Sql.Test.ScenarioTests.ManagedInstanceActiveDirectoryAdministratorTests.TestManagedInstanceActiveDirectoryAdministrator() in D:\a\1\s\src\Sql\Sql.Test\ScenarioTests\ManagedInstance
> > 
> > 
> > @msJinLei @dingmeng-xue any idea about this. looks like tests scenarios not working with service principals.
> 
> In your updated code, there are API calls to get serviceprincipal, so please re-record testcase: https://github.com/Azure/azure-powershell/blob/amagarwa/dev/sqlmiaddSPGuestuserSupport/src/Sql/Sql.Test/ScenarioTests/ManagedInstanceActiveDirectoryAdministratorTests.ps1

I see. that is the difficult part, last time I tried for 3 days and could not get it to work. Last dev has lef the team. :(, let me check if we want to re-record or just disable the test until we has a resource to fix test.

**Comment by amolagar5** [comment]
> * Updated changelog.md
>   please update ## Upcoming release session only
> * please add test cases to cover the new added logic

I have marked the test as skip.  I debugged more and looks like its the token that 's going with graph call, its wrong and using https://management.core.windows.net/ and not https://graph.windows.net/, looks like in debugging all places I added my token are called fine. so puzzling. When calling ActiveDirectoryClient.FilterGroups(filter).Where(gr => string.Equals(gr.DisplayName, displayName, StringComparison.OrdinalIgnoreCase)); in MockScenario, it picks up wrong token. Wondering if problem is in ActiveDirectoryClient itself.

**Pull request merged**